hudi-spark-datasource/hudi-spark/src/test/java/org/apache/hudi/functional/TestBootstrapRead.java

```diff
@@ -0,0 +1,297 @@
+import org.apache.hudi.client.bootstrap.selector.MetadataOnlyBootstrapModeSelector;
+import org.apache.hudi.common.testutils.HoodieTestDataGenerator;
+import org.apache.hudi.config.HoodieBootstrapConfig;
+import org.apache.hudi.config.HoodieCompactionConfig;
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.keygen.ComplexKeyGenerator;
+import org.apache.hudi.keygen.NonpartitionedKeyGenerator;
+import org.apache.hudi.keygen.SimpleKeyGenerator;
+import org.apache.hudi.testutils.HoodieSparkClientTestBase;
+
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SaveMode;
+import org.apache.spark.sql.functions;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.io.TempDir;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.apache.hudi.common.testutils.RawTripTestPayload.recordToString;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+/**
+ * Tests different layouts for bootstrap base path
+ */
+@Tag("functional")
+public class TestBootstrapRead extends HoodieSparkClientTestBase {
+
+  @TempDir
+  public java.nio.file.Path tmpFolder;
+  protected String bootstrapBasePath = null;
+  protected String bootstrapTargetPath = null;
+  protected String hudiBasePath = null;
+
+  protected static int nInserts = 100;
+  protected static int nUpdates = 20;
+  protected static String[] dashPartitionPaths = {"2016-03-15", "2015-03-16", "2015-03-17"};
+  protected static String[] slashPartitionPaths = {"2016/03/15", "2015/03/16", "2015/03/17"};
+  protected String bootstrapType;
+  protected Boolean dashPartitions;
+  protected String tableType;
+  protected Integer nPartitions;
+
+  protected String[] partitionCols;
+  protected static String[] dropColumns = {"_hoodie_commit_time", "_hoodie_commit_seqno", "_hoodie_record_key",  "_hoodie_file_name", "city_to_state"};
+
+  @BeforeEach
+  public void setUp() throws Exception {
+    bootstrapBasePath = tmpFolder.toAbsolutePath() + "/bootstrapBasePath";
+    hudiBasePath = tmpFolder.toAbsolutePath() + "/hudiBasePath";
+    bootstrapTargetPath = tmpFolder.toAbsolutePath() + "/bootstrapTargetPath";
+    initSparkContexts();
+  }
+
+  @AfterEach
+  public void tearDown() throws IOException {
+    cleanupSparkContexts();
+    cleanupClients();
+    cleanupTestDataGenerator();
+  }
+
+  private static Stream<Arguments> testArgs() {
+    Stream.Builder<Arguments> b = Stream.builder();
+    //TODO: add dash partitions false with [HUDI-4944]
```

Comment:
please add more details here.
Also, you could try adding a sleep of 2s and check if timeline server returns correct fs view.

Reply:
1. fixed
2.  I added a 5 second delay between bootstrapping and the update and was still getting failures at the same rate

Reply:
Ok let's tackle that issue separately but I don't see more details in the comment.

Reply:
I changed the line below to:
Boolean[] dashPartitions = {true/*,false*/};

The code is already there to test that, it just doesn't work currently because of the double decoding issue